src/JustEat.StatsD/V2/StatsDPublisherV2.cs

```diff
@@ -0,0 +1,170 @@
+            {
+                var msg = StatsDMessage.Counter(value, bucket);
+                SendMessage(sampleRate, msg);
+            }
+        }
+
+        public void Increment(long value, double sampleRate, params string[] buckets)
+        {
+            foreach (var bucket in buckets)
+            {
+                Increment(value, sampleRate, bucket);
+            }
+        }
+
+        public void Decrement(string bucket)
+        {
+            Increment(-1, DefaultSampleRate, bucket);
+        }
+
+        public void Decrement(long value, string bucket)
+        {
+            Increment(value > 0 ? -value : value, DefaultSampleRate, bucket);
+        }
+
+        public void Decrement(long value, double sampleRate, string bucket)
+        {
+            Increment(value > 0 ? -value : value, sampleRate, bucket);
+        }
+
+        public void Decrement(long value, double sampleRate, params string[] buckets)
+        {
+            foreach (var bucket in buckets)
+            {
+                Increment(value > 0 ? -value : value, sampleRate, bucket);
+            }
+        }
+
+        public void Gauge(double value, string bucket)
+        {
+            SendMessage(DefaultSampleRate, StatsDMessage.Gauge(value, bucket));
+        }
+
+        public void Gauge(double value, string bucket, DateTime timestamp)
+        {
+            Gauge(value, bucket);
+        }
+
+        public void Gauge(long value, string bucket)
+        {
+            Gauge((double) value, bucket);
+        }
+
+        public void Gauge(long value, string bucket, DateTime timestamp)
+        {
+            Gauge(value, bucket);
+        }
+
+        public void Timing(TimeSpan duration, string bucket)
+        {
+            Timing(duration.Ticks, bucket);
+        }
+
+        public void Timing(TimeSpan duration, double sampleRate, string bucket)
+        {
+            Timing(duration.Ticks, sampleRate, bucket);
+        }
+
+        public void Timing(long duration, string bucket)
+        {
+            Timing(duration, DefaultSampleRate, bucket);
+        }
+
+        public void Timing(long duration, double sampleRate, string bucket)
+        {
+            var msg = StatsDMessage.Timing(duration, bucket);
+            SendMessage(sampleRate, msg);
+        }
+
+        public void MarkEvent(string name)
+        {
+            Increment(name);
+        }
+
+        private void SendMessage(double sampleRate, StatsDMessage msg)
+        {
+            var buffer = Buffer();
+
+            if (_formatter.TryFormat(msg, sampleRate, buffer, out int written))
+            {
+                _transport.Send(new ArraySegment<byte>(buffer, 0, written));
+            }
+            else
+            {
+                var newSize = _formatter.GetBufferSize(msg);
+                var size = Interlocked.CompareExchange(ref _bufferSize, buffer.Length, newSize);
+                _buffer = new byte[size];
```

Comment:
It's a bit weird here that we change `_buffer` and then call it through `Buffer()`, rather than just use it.